R/MsBackendTimsTof-functions.R

```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
         if (!length(sd <- setdiff(columns, c("frame", "scan"))))
             stop("At least one column value different from 'frame' and",
                  " 'scan' required")
-        tmp <- query(tms, unique(x@indices[I, "frame"]), c("frame", "scan", sd))
-        f <- factor(paste(tmp$frame, tmp$scan),
-                    levels = paste(x@indices[I, "frame"], x@indices[I, "scan"]))
-        if (length(sd) == 1)
-            res[I] <- unname(split(tmp[, sd], f))
-        else res[I] <- unname(split.data.frame(as.matrix(tmp[, sd]), f))
+        ## Ensure mz and intensity are always first, if provided
+        sd <- c(intersect(c("mz", "intensity"), sd),
+                setdiff(sd, c("mz", "intensity")))
+        i_frame <- x@indices[I, "frame"]
+        tmp <- query(tms, unique(i_frame), c("frame", "scan", sd))
+        ## subset tmp if we're about to extract only few scans.
+        if (length(i_frame) < (nrow(tmp) / 10)) {
+            i_scan <- x@indices[I, "scan"]
+            tmp <- tmp[tmp$scan %in% i_scan, ]
+            rownames(tmp) <- NULL
+            ids <- paste(i_frame, i_scan)
+        } else ids <- paste(i_frame, x@indices[I, "scan"])
+        if (!nrow(tmp))
```

Comment:
This is to avoid a problem if no peaks was recorded for **any** of the requested scans (spectra): without that a `list` with `logical` matrices is returned (instead of `numeric`).